From 16cd0fbec96e0bbe87c8c207054ff307d06fd767 Mon Sep 17 00:00:00 2001
From: Igal Zeltser <izeltser@habana.ai>
Date: Thu, 28 Mar 2024 15:02:07 +0200
Subject: [PATCH] [FSW-334] fs1: pci: fix bar size setting

Bar size is set by writing a mask to the shadow register, not the actual
DBI register. the shadow register defines which bits can't be written by
the host, and that is how the host determines the size during
enumeration.
At this stage only Bar0 is resized to 1GB, the other BARs are remain
with default size.

Change-Id: I70620d707bc6592d3571bf267f51a36fac114f33
Signed-off-by: Igal Zeltser <izeltser@habana.ai>
---

diff --git a/arch/fs1/arc/pldm.c b/arch/fs1/arc/pldm.c
index d9ca7db..9d42711 100644
--- a/arch/fs1/arc/pldm.c
+++ b/arch/fs1/arc/pldm.c
@@ -154,8 +154,7 @@
 	}
 }
 
-#define BAR_TYPE_MEM	0xC
-#define BAR_SIZE_1GB	(((SZ_1G - 1) & PCIE_DBI_BAR0_REG_BAR0_START_M) | BAR_TYPE_MEM)
+#define BAR_SIZE_1GB	(SZ_1G - 1)
 
 static void pldm_pcie_params_init(void)
 {
@@ -166,6 +165,9 @@
 	val |= PCIE_DBI_MISC_CONTROL_1_OFF_DBI_RO_WR_EN_M;
 	regmap_writel(val, IO0_H0_PCIE_DBI(mmPCIE_DBI_MISC_CONTROL_1_OFF));
 
+	/* switch to shadow registers access in order to set BAR masks */
+	regmap_writel(PCIE_AUX_DBI_CS2_M, mmIO_0_CFG_HALF_0_CXL_EP_PCIE_AUX_BASE + mmPCIE_AUX_DBI);
+
 	/* set PF CFG BAR to 1GB */
 	regmap_writel(BAR_SIZE_1GB, IO0_H0_PCIE_DBI(mmPCIE_DBI_BAR0_REG));
 	regmap_writel(0, IO0_H0_PCIE_DBI(mmPCIE_DBI_BAR1_REG));
@@ -174,6 +176,9 @@
 	regmap_writel(BAR_SIZE_1GB, IO0_H0_PCIE_DBI(mmPCIE_DBI_SRIOV_BAR0_REG));
 	regmap_writel(0, IO0_H0_PCIE_DBI(mmPCIE_DBI_SRIOV_BAR1_REG));
 
+	/* disable shadow registers access */
+	regmap_writel(0, mmIO_0_CFG_HALF_0_CXL_EP_PCIE_AUX_BASE + mmPCIE_AUX_DBI);
+
 	pldm_pcie_dynamic_init();
 
 	/* disable writing to DBI RO regs */
